Mathlib/ModelTheory/Substructures.lean

```diff
@@ -967,6 +983,52 @@ theorem range_subtype (S : L.Substructure M) : S.subtype.toHom.range = S := by
 lemma subtype_comp_inclusion {S T : L.Substructure M} (h : S ≤ T) :
     T.subtype.comp (inclusion h) = S.subtype := rfl
 
+namespace Equiv
+
+/-- Equivalence between equal substructures. -/
+def ofEq {S T : L.Substructure M} (h : S = T) : S ≃[L] T :=
+  { Equiv.setCongr (congr_arg _ h) with }
+
+@[simp]
+theorem ofEq_refl (S : L.Substructure M) : (ofEq (Eq.refl S)) = Equiv.refl L S := rfl
+
+@[simp]
+theorem symm_ofEq {S T : L.Substructure M} (h : S = T) :
+    (ofEq h).symm = ofEq h.symm := rfl
+
+@[simp]
+theorem ofEq_comp_ofEq {S T U : L.Substructure M} (h : S = T) (h' : T = U) :
+    (ofEq h').comp (ofEq h) = ofEq (h.trans h') := rfl
+
+@[simp]
+theorem ofEq_apply {S T : L.Substructure M} (h : S = T) {m : M} (hm : m ∈ S) :
+    ofEq h ⟨m, hm⟩ = ⟨m, h ▸ hm⟩ := rfl
+
+@[simp]
+theorem subtype_comp_EquivOfEq {S T : L.Substructure M} (h : S = T) :
+    T.subtype.comp (ofEq h).toEmbedding = S.subtype := rfl
+
+@[simp]
+theorem substructureEquivMap_refl (S : L.Substructure M) :
+    (Embedding.refl L M).substructureEquivMap S = ofEq (map_id S).symm := by
+  ext ⟨⟩
+  simp only [Embedding.refl_toHom]
+  rfl
+
+@[simp]
+theorem substructureEquivMap_comp_substructureEquivMap (f : M ↪[L] N) (g : N ↪[L] P)
+    (S : L.Substructure M) : (g.substructureEquivMap (S.map f.toHom)).comp
+    (f.substructureEquivMap S) = (ofEq (S.map_map ..).symm).comp
+    ((g.comp f).substructureEquivMap S) := by
+  apply Equiv.injective_toEmbedding
+  apply (subtype _).comp_injective
+  simp only [Equiv.comp_toEmbedding, ← Embedding.comp_assoc, Embedding.subtype_substructureEquivMap,
+    subtype_comp_EquivOfEq]
+  ext
+  rfl
```

Comment:
This statement is currently very hard to read. In general, don't start multiline return types in the middle of a line.
```suggestion
    (S : L.Substructure M) :
    (g.substructureEquivMap (S.map f.toHom)).comp (f.substructureEquivMap S) =
      (ofEq (S.map_map ..).symm).comp ((g.comp f).substructureEquivMap S) := by aesop
```